### PR TITLE
Custom resource rewrite

### PR DESCRIPTION
**Pull request by rshade**
This is the pull request for #151. It's nowhere near ready yet. :)

### CODE REVIEW

**Comment by damacus** [comment]
Resources
- [x] frontend
- [x] backend
- [x] global_config
- [x] default_config
- [x] package install 
- [x] source install 

@teknofire and the others I've tagged for review. It'd be great to hear some feedback on the direction we're heading with this rewrite, and to see if it meets your needs, at least in some generic way. 

Thanks!

File: libraries/matchers.rb [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
-  end
-
-  def create_haproxy_lb(name)
-    ChefSpec::Matchers::ResourceMatcher.new(:haproxy_lb, :create, name)
-  end
-
-  def create_haproxy_config(name)
-    ChefSpec::Matchers::ResourceMatcher.new(:haproxy_config, :create, name)
-  end
+  # def delete_haproxy(name)
```
Comment (martinisoft) [inline libraries/matchers.rb]: Any reason for all this commented code?

File: libraries/matchers.rb [hunk 2]
```diff
@@ -1,26 +1,26 @@
 if defined?(ChefSpec)
 
-  ChefSpec.define_matcher :haproxy
-  ChefSpec.define_matcher :haproxy_lb
-  ChefSpec.define_matcher :haproxy_config
+  ChefSpec.define_matcher :haproxy_haproxy
```
Comment (martinisoft) [inline libraries/matchers.rb]: This looks really, really weird to me. Any way to not do this?

File: metadata.rb [hunk 3]
```diff
@@ -4,19 +4,16 @@
...
 description       'Installs and configures haproxy'
 long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
-version           '3.0.1'
-
-recipe 'haproxy', 'Installs and configures haproxy'
-recipe 'haproxy::app_lb', 'Installs and configures haproxy by searching for nodes of a particular role'
+version           '4.0.0'
+source_url        'https://github.com/sous-chefs/haproxy' if respond_to?(:source_url)
+issues_url        'https://github.com/sous-chefs/haproxy/issues' if respond_to?(:issues_url)
+chef_version      '>= 12.5' if respond_to?(:chef_version)
```
Comment (martinisoft) [inline metadata.rb]: You can probably avoid the respond_to portion since you're aiming to be a 12.5+ compatible cookbook.

File: metadata.rb [hunk 4]
```diff
@@ -4,19 +4,16 @@
...
 
 %w( debian ubuntu centos redhat scientific oracle ).each do |os|
   supports os
 end
 
+depends           'compat_resource', '>= 12.16.3'
```
Comment (martinisoft) [inline metadata.rb]: This dependency can go away since you're aiming to be 12.5+

File: test/fixtures/cookbooks/test/recipes/config_1.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+haproxy 'package'
+
+haproxy_config_global '' do
```
Comment (martinisoft) [inline test/fixtures/cookbooks/test/recipes/config_1.rb]: I'd wonder if there is a better way to scope this.

File: test/integration/source/source_spec.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (martinisoft) [inline test/integration/source/source_spec.rb]: Don't need this extra space at the top.

File: libraries/matchers.rb (hunk 2 above)
Comment (damacus) [inline libraries/matchers.rb]: Yeah, that bit doesn't match, I couldn't get it matching with  `:haproxy` lord knows why.

File: test/fixtures/cookbooks/test/recipes/config_1.rb (hunk 5 above)
Comment (damacus) [inline test/fixtures/cookbooks/test/recipes/config_1.rb]: I've aimed at slicing it into config sections. All ears if there's something better.

File: resources/config_defaults.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
+property :timeout, Hash, default: { client: '10s', server: '10s', connect: '10s' }
+property :name, String, name_property: true
+property :log, String, default: 'global'
+property :mode, String, default: 'http', equal_to: %w(http tcp)
+property :balance, default: 'roundrobin', equal_to: %w(roundrobin static-rr leastconn first source uri url_param header rdp-cookie)
+property :option, Array, default: %w(httplog dontlognull redispatch)
+property :status_uri, String, default: '/haproxy-status'
```
Comment (teknofire) [inline resources/config_defaults.rb]: This should probably be `stats_uri` to match the property from the listen resource and the config option name in `haproxy`

Comment (teknofire) [inline resources/config_defaults.rb]: Also this doesn't appear to get used at all by the `haproxy.cfg.erb`

**Comment by damacus** [comment]
Closes #151, #111, 
Closes #159 ACLs are a Hash/Array inside a backends & frontends
Closes #148 we no longer use Chef Search.

Possibly cleans up #58 as we have a clean way of defining backends.

File: templates/default/haproxy.cfg.erb [hunk 8]
```diff
@@ -1,40 +1,149 @@
...
+frontend <%= frontend %>
+  default_backend <%= f['default_backend'] %>
+  <% unless f['bind'].empty? -%>bind <%= f['bind'] %><% end -%>
+  <% unless f['maxconn'].empty? -%>maxconn <%= f['maxconn'] %><% end -%>
+  <% unless f['http_request'].empty? -%>http_request<%= f['http_request'] %><% end -%>
+  <% unless f['http_response'].empty? -%>http_response<%= f['http_response'] %><% end -%>
+
+  <% unless f['use_backend'].nil? %>
+  <% f['use_backend'].each do | backend |%>
+  <% backend.each do | backend | %>
```
Comment (teknofire) [inline templates/default/haproxy.cfg.erb]: Having the double each here seems odd to me, should this be `f['use_backend'].flatten.each do |backend|`

**Pull request merged**